subsys/net/lib/lwm2m/lwm2m_obj_swmgmt.c

```diff
@@ -0,0 +1,743 @@
+			instance->update_state = UPD_STATE_INITIAL;
+			instance->update_result = UPD_RES_INITIAL;
+		case EVENT_DELETE_PACKAGE:
+			ret = instance->delete_package_cb(instance->obj_inst_id);
+			if (ret == 0) {
+				instance->update_state = UPD_STATE_INITIAL;
+				instance->update_result = UPD_RES_INITIAL;
+			}
+			break;
+		default:
+			ret = -EINVAL;
+			break;
+		}
+		break;
+	default:
+		ret = -EINVAL;
+		break;
+	}
+
+	return ret;
+}
+
+static int install_cb(uint16_t obj_inst_id)
+{
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+
+	return handle_event(instance, EVENT_INSTALL);
+}
+
+int lwm2m_swmgmt_install_completed(uint16_t obj_inst_id, int error_code)
+{
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+
+	if (error_code) {
+		error_code = handle_event(instance, EVENT_INSTALL_FAIL);
+	} else {
+		error_code = handle_event(instance, EVENT_INSTALL_SUCCESSFUL);
+	}
+
+	return error_code;
+}
+
+static int uninstall_cb(uint16_t obj_inst_id)
+{
+	/*
+	 * TODO Implement handling for "ForUpdate".
+	 * This requires the Zephyr LWM2M to handle arguments to exec calls.
+         */
+
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+	return handle_event(instance, EVENT_DELETE_PACKAGE);
+}
+
+static int activate_cb(uint16_t obj_inst_id)
+{
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+	return handle_event(instance, EVENT_ACTIVATE);
+}
+
+static int deactivate_cb(uint16_t obj_inst_id)
+{
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+	return handle_event(instance, EVENT_DEACTIVATE);
+}
+
+static int package_write_cb(uint16_t obj_inst_id, uint16_t res_id, uint16_t res_inst_id,
+			    uint8_t *data, uint16_t data_len, bool last_block, size_t total_size)
+{
+	int ret = -EINVAL;
+	struct lwm2m_swmgmt_data *instance = NULL;
+	instance = find_index(obj_inst_id);
+
+
+	ret = handle_event(instance, EVENT_PKG_URI_WRITE);
+
+	if (ret < 0) {
+		return ret;
+	}
+
+	ret = instance->write_package_cb(obj_inst_id, res_id, res_inst_id, data, data_len,
+					 last_block, total_size);
+
+	if (ret < 0) {
+		handle_event(instance, EVENT_DOWNLOAD_FAILED);
+		switch (ret) {
+		case -ENOMEM:
+			instance->update_result = UPD_RES_OUT_OF_MEMORY_DURING_DOWNLOAD;
+			break;
+		case -ENOSPC:
+			instance->update_result = UPD_RES_NOT_ENOUGH_STORAGE;
+			ret = -EFBIG;
+			break;
+		case -EFAULT:
+			instance->update_result = UPD_RES_PACKAGE_INTEGRITY_CHECK_FAILURE;
+			break;
+		default:
+			instance->update_result = UPD_RES_LOST_CONNECTION_DURING_DOWNLOAD;
+			break;
+		}
+
+		return ret;
+	}
+
+	if (last_block) {
+		handle_event(instance, EVENT_PKG_WRITTEN);
+	}
+
+	return 0;
+}
+
+static void set_update_result_from_error(struct firmware_pull_context *context, int error_code)
+{
+	struct lwm2m_swmgmt_data *instance;
+	instance = CONTAINER_OF(context, struct lwm2m_swmgmt_data, pull_context);
```

Comment:
There isn't `firmware_pull_context` in `lwm2m_swmgmt_data` when `CONFIG_LWM2M_FIRMWARE_UPDATE_PULL_SUPPORT` isn't enabled. Is this supposed only to be used with the PULL-method? What about PUSH-method?

Reply:
This doesn't support PUSH atm. This is something that should be implemented in the future. Any safety check that we could introduce here?

Reply:
That's a bit unfortunate to be honest. Do  Iunderstand correctly, that the only limitation is the PULL context provided to the result callback? Perhaps instead of providing this pointer, we could simple feed the callback with the instance number? On the FW update object side it's not used anyway, and here we only use the pointer to obtain the acual object instance - we could achieve the same with instance number and `find_index()`

Reply:
Lack of PUSH is indeed unfortunate. Any estimates that how long it would take to implement?

Reply:
Since I've inherited this PR I cannot answer how long it would take, right now. I'm just not knowledgeable enough about this topic. I'll do my best to do my homework and see if I can implement it in a reasonable time.

Reply:
I thought about this and I think I should not implement the PUSH method. That's because I have no way to test it in the foreseeable future. I wouldn't feel comfortable with merging this without testing it. I might be the one who implements that in the future but I can't do it right now.

Reply:
Since PUSH means just pushing data into the Software Management object, this should be supported.